substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    public static boolean isUsedThisEpoch(Class<?> clazz) {
+        return predicate(clazz, TRANSIENT_BIT | JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    public static long getTraceIdRaw(Class<?> clazz) {
+        JfrTraceIdMap map = getTraceIdMap();
+        assert map != null;
+        return getTraceIdMap().getId(clazz);
+    }
+
+    public static long getTraceId(Class<?> clazz) {
+        long traceid = getTraceIdRaw(clazz);
+        return traceid >>> TRACE_ID_SHIFT;
+    }
+
+    public static long load(Class<?> clazz) {
+        return JfrTraceIdLoadBarrier.load(clazz);
+    }
+
+    private static void tagAsJdkJfrEvent(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        assert id != -1;
+        map.setId(index, id | JDK_JFR_EVENT_KLASS);
+    }
+
+    private static void tagAsJdkJfrEventSub(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        assert id != -1;
+        map.setId(index, id | JDK_JFR_EVENT_SUBKLASS);
+    }
+
+    private static boolean isEventClass(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        assert id != -1;
+        return (id & (JDK_JFR_EVENT_KLASS | JDK_JFR_EVENT_SUBKLASS)) != 0;
+    }
+
+    private static boolean setSystemEventClass(Class<?> clazz, int index) {
+        if ((clazz == jdk.internal.event.Event.class || clazz == jdk.jfr.Event.class) &&
+            (clazz.getClassLoader() == null || clazz.getClassLoader() == ClassLoader.getSystemClassLoader())) {
+            tagAsJdkJfrEvent(index);
+            return true;
+        }
+        return false;
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public static void assign(Class<?> clazz, Map<Class<?>, Integer> classToIndex) {
+        assert clazz != null;
+        int index = classToIndex.get(clazz) + 1; // Off-set by one for error-catcher
+        if (getTraceIdMap().getId(index) != -1) {
+            return;
+        }
+        long typeId = JVM.getJVM().getTypeId(clazz);
+        getTraceIdMap().setId(index, typeId << TRACE_ID_SHIFT);
+        if (!setSystemEventClass(clazz, index)) {
+            Class<?> superClazz = clazz.getSuperclass();
+            if (superClazz != null) {
+                int superIndex = classToIndex.get(superClazz) + 1;
+                if (getTraceIdMap().getId(superIndex) != -1) {
+                    assign(superClazz, classToIndex);
+                }
+                if (isEventClass(superIndex)) {
+                    tagAsJdkJfrEventSub(index);
+                }
+            }
+        }
+    }
+
+    public static boolean isSerialized(Class<?> clazz) {
+        return predicate(clazz, SERIALIZED_BIT);
+    }
+
+    public static void setSerialized(Class<?> clazz) {
+        long id = getTraceIdMap().getId(clazz);
+        getTraceIdMap().setId(clazz, id | SERIALIZED_BIT);
```

Comment:
This pattern of reading the trace id and then modifying it can cause concurrency issues. I suppose this works because `setSerialized(...)` and `clearSerialized(...)` would only be called at a safepoint? If so, please add an assertion.